File: cityzoom/DataLayer/DLAPI/src/main/java/API/Routes/Devices.java
package API.Routes;

import API.Aux.MongoAux;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.mongodb.client.FindIterable;
import org.bson.Document;
import org.json.JSONException;
import spark.Request;
import spark.Response;

import javax.net.ssl.HttpsURLConnection;
import java.util.*;

import static API.DataLayerAPI.*;
import static com.mongodb.client.model.Filters.*;

public class Devices {
    private static Date date = new Date();

    // Status - Passing
    public static String createDevice(Request request, Response response) throws JSONException {
        logger.info("Creating Device Request");
        response.type("application/json");
        JsonObject body = (JsonObject) MongoAux.jsonParser.parse(request.body());
        String valid = validator.validateCreateDevice(body);
        if (!valid.equals("")) {
            logger.error("Error Creating Device");
            response.status(HttpsURLConnection.HTTP_BAD_REQUEST);
            return valid;
        }
        Set<String> keys = body.keySet();
        String device_name = body.get("device_name").getAsString();
        String provider = body.get("provider").getAsString();
        JsonArray vertical = body.get("vertical").getAsJsonArray();
        boolean mobile = body.get("mobile").getAsBoolean();
        double latitude = 0;
        double longitude = 0;
        if (!mobile) {
            latitude = body.get("latitude").getAsDouble();
            longitude = body.get("longitude").getAsDouble();
        }
        /*if (!types.contains(type)){
            response.status(HttpsURLConnection.HTTP_CONFLICT);
            return "{\n" +
                    "\t\"status\": \"Error @ type\",\n" +
                    "\t\"Error\": \"Type " + type + " not available. Available types are "+types.toString()+"\"\n" +
                    "}";

        }*/
        String description = keys.contains("description") ? body.get("description").getAsString() : "";
        String municipal = keys.contains("municipality") ? body.get("municipality").getAsString() : "";
        String produceRequest =
                "{\n" +
                        "\t\"device_name\":\""+device_name+"\",\n" +
                        "\t\"description\":\""+description+"\",\n" +
                        "\t\"provider\":\""+provider+"\",\n" +
                        "\t\"vertical\":"+vertical+",\n" +
                        "\t\"mobile\":"+mobile+",\n" +
                        "\t\"municipality\":\""+municipal+"\",\n" +
                        "\t\"latitude\": "+latitude+",\n"+
                        "\t\"longitude\": "+longitude+",\n"+
                        "\t\"creation\": "+date.getTime()+"\n" +
                        "}";
        String topic = "Devices";
        producer.produce(topic, "chave_minima", produceRequest);
        logger.info("Device sent to kafka");
        return "{\n" +
                "\t\"status\":\"Creation successful\",\n" +
                "\t\"device_name\":\""+device_name+"\",\n" +
                "\t\"account_name\":\"none\",\n" +
                "\t\"creation_time\":\""+date.getTime()+"\",\n" +
                "\t\"mobile\":\""+mobile+"\"\n" +
                "}";
    }

    // Status - Passing
    public static String deleteDevice(Request request, Response response) {
        logger.info("Deleting device");
        response.type("application/json");
        String to_del = request.params(":device_id");
        if (devices.find(eq("device_name", to_del)).first() == null) {
            response.status(HttpsURLConnection.HTTP_NOT_FOUND);
            return "{\n" +
                    "\t\"status\": \"Error\",\n" +
                    "\t\"Error\": \"Stream " + to_del + " not found.\"\n" +
                    "}";
        }
        devices.deleteOne(eq("device_id", to_del));
        FindIterable<Document> streamsToDel = streams.find(eq("device_id", to_del));
        streams.deleteMany(eq("device_id", to_del));
        for (Document d : streamsToDel) {
            values.deleteMany(eq("stream_name", d.getString("stream")));
        }
        response.status(HttpsURLConnection.HTTP_NO_CONTENT);
        return "";
    }

    // Status - Passing
    public static String listDevices(Request request, Response response) {
        logger.info("Listing all devices");
        response.type("application/json");
        long start = request.queryParams("interval_start") != null ?
                Long.parseLong(request.queryParams("interval_start")) : 0;
        long end = request.queryParams("interval_end") != null?
                Long.parseLong(request.queryParams("interval_end")) : date.getTime();
        long compass = date.getTime();

        if (end < start || end > compass || start < 0) {
            logger.error("Bad time interval");
            response.status(HttpsURLConnection.HTTP_NOT_ACCEPTABLE);
            return "{\n" +
                    "\t\"Error\": \"Interval not acceptable\"\n" +
                    "}";
        }

        List<String> userDevices = new ArrayList<>();
        long size = devices.countDocuments(and(gte("creation", start), lte("creation", end)));
        long total = devices.countDocuments();
        FindIterable<Document> streamsIterable = devices.find(and(gte("creation", start), lte("creation", end)));
        JsonObject jsonDev;
        for (Document document : streamsIterable) {
            jsonDev = (JsonObject) MongoAux.jsonParser.parse(document.toJson());

            FindIterable<Document> streamDocs = streams.find(eq("device_id", jsonDev.get("_id").getAsJsonObject().get("$oid").getAsString()));
            ArrayList<String> streamList = new ArrayList<>();
            for (Document doc : streamDocs) {
                JsonObject jsonStream = (JsonObject) MongoAux.jsonParser.parse(doc.toJson());
                streamList.add(jsonStream.get("_id").getAsJsonObject().get("$oid").getAsString());
            }

            System.out.println(jsonDev.toString());
            String device =
                    "{\n" +
                            "\t\"device_id\": \""+jsonDev.get("_id").getAsJsonObject().get("$oid").getAsString()+"\",\n" +
                            "\t\"mobile\": "+jsonDev.get("mobile").getAsBoolean()+",\n" +
                            "\t\"municipality\": \""+jsonDev.get("municipality").getAsString()+"\",\n" +
                            "\t\"vertical\": "+jsonDev.get("vertical").getAsJsonArray()+",\n" +
                            "\t\"provider\": \""+jsonDev.get("provider").getAsString()+"\",\n" +
                            "\t\"streams\": "+ Arrays.toString(streamList.toArray()) +",\n" +
                            "\t\"device_name\": \""+jsonDev.get("device_name").getAsString()+"\",\n" +
                            "\t\"description\": \""+jsonDev.get("description").getAsString()+"\",\n" +
                            "\t\"locations\": "+jsonDev.get("locations").getAsJsonArray()+",\n"+
                            "\t\"creation\": "+jsonDev.get("creation").getAsJsonObject().get("$numberLong").getAsLong()+"\n" +
                            "}";
            userDevices.add(device);
        }
        response.status(HttpsURLConnection.HTTP_OK);
        if (end != compass || start != 0) {
            return "{\n" +
                    "\t\"total_devices\": "+total+",\n" +
                    "\t\"size\": "+size+",\n" +
                    "\t\"start\": "+start+",\n" +
                    "\t\"end\": "+end+",\n" +
                    "\t\"user_devices\": "+userDevices.toString()+"\n" +
                    "}";
        }
        return "{\n" +
                "\t\"total_devices\": "+total+",\n" +
                "\t\"user_devices\": "+userDevices.toString()+"\n" +
                "}";
    }

    // Status - Passing
    public static String detailDevice(Request request, Response response) {
        logger.info("Detailing a device");
        response.type("application/json");
        String device = request.params(":device_id");
        Document document = devices.find(eq("device_name",device)).first();
        if (document == null) {
            logger.error("Device not found");
            response.status(HttpsURLConnection.HTTP_NOT_FOUND);
            return "{\n" +
                    "\t\"status\": \"Error\",\n" +
                    "\t\"Error\": \"Device " + device + " not found.\"\n" +
                    "}";
        }
        FindIterable<Document> streamDocs = streams.find(eq("device_id", device));
        ArrayList<String> streamList = new ArrayList<>();
        for (Document doc : streamDocs) {
            JsonObject jsonStream = (JsonObject) MongoAux.jsonParser.parse(doc.toJson());
            streamList.add(jsonStream.get("_id").getAsJsonObject().get("$oid").getAsString());
        }
        response.status(200);
        JsonObject jsonDev = (JsonObject) MongoAux.jsonParser.parse(document.toJson());
        return "{\n" +
                "\t\"device_id\": \""+jsonDev.get("_id").getAsJsonObject().get("$oid").getAsString()+"\",\n" +
                "\t\"mobile\": "+jsonDev.get("mobile").getAsBoolean()+",\n" +
                "\t\"municipality\": \""+jsonDev.get("municipality").getAsString()+"\",\n" +
                "\t\"vertical\": "+jsonDev.get("vertical").getAsJsonArray()+",\n" +
                "\t\"provider\": \""+jsonDev.get("provider").getAsString()+"\",\n" +
                "\t\"streams\": "+ Arrays.toString(streamList.toArray()) +",\n" +
                "\t\"device_name\": \""+jsonDev.get("device_name").getAsString()+"\",\n" +
                "\t\"description\": \""+jsonDev.get("description").getAsString()+"\",\n" +
                "\t\"locations\": "+jsonDev.get("locations").getAsJsonArray()+",\n"+
                "\t\"creation\": "+jsonDev.get("creation").getAsJsonObject().get("$numberLong").getAsLong()+"\n" +
                "}";
    }
}

File: cityzoom/data_layer/src/routes/devices.js
const express = require('express')
const {validateCreateDevice} = require('../validation')
const { validation } = require('../middleware')
const devicesDebug = require('debug')('app:Devices')
const devices = require('../db/models/devices')
const values = require('../db/models/values')
//Broker producer and consumer
const producer = require('../producer')

const router = new express.Router()

// create a device
router.post('', validation(validateCreateDevice, 'body', 'Invalid device'), async (req, res) => {
    // convert request to broker-stuff
    const to_broker = {
        ...req.body,
        created_at: Number(Date.now()), 
        location: []
    }
    
    //Publishes the device in the broker
    const wasPublished = await producer.publish('cityzoom/devices',to_broker)
    if(!wasPublished){
        devicesDebug(`Device ${to_broker.device_ID} already exists`)
        return res.status(409).send({'Error':`Device ${to_broker.device_ID} already exists`}) 
    }

    return res.status(201).send({ 
        status: 'Creation successful',
        device_ID: req.body['device_ID'],
        device_name: req.body['device_name'],
        mobile: req.body['mobile'],
        provider: req.body['provider'],
        created_at: Number(Date.now())
    })
})

// get all devices
router.get('', async (req, res) => {
    devicesDebug('[DEBUG] Fetching all Devices')
    const start = req.query.interval_start ? req.query.interval_start : 0
    const compass = Number(Date.now())
    const end = req.query.interval_end ? req.query.interval_end : compass
    if (end < start || start < 0) {
        devicesDebug('[ERROR] Interval is wrong')
        return res.status(400).send({error: 'Bad interval defined'})
    }
    var allDevices = await devices.find({created_at: { $gte: start, $lte: end}})
    res.status(200).send(allDevices)
})

router.get("/location", async(req,res) => {
    devicesDebug('[DEBUG] Fetching last device locations')
    const tmp = await values.aggregate([{
        $group:{
            _id: {
              device_id: "$device_ID",
              longitude: "$longitude",
              latitude: "$latitude"
            },
            last: {
              $max: "$timestamp"
            }
          }
    }])
    console.log(tmp)
    res.send(tmp)
})

// get device by ID
router.get('/:id', async (req, res) => {
    const doc = await devices.findOne({device_ID:req.params.id})
    if (!doc) { return res.status(404).send({'Status':'Not Found'}) }

    res.status(200).send(doc)
})

// delete device by ID
router.delete('/:id', async (req, res) => {
    const deletion = await devices.deleteOne({device_ID:req.params.id})
    if (deletion.deletedCount == 0) { return res.status(404).send({'Status':'Not Found'})}
    res.status(204).send()
})

// get values by stream
router.get('/:id/values', async (req,res) => {
    const device = await devices.findOne({device_ID:req.params.id})

    console.log(device)
    if (!device) { return res.status(404).send({'Status':'Not found'})}
    var start = req.query.interval_start ? Number(req.query.interval_start) : Number(new Date(0))
    var end = req.query.interval_end ? Number(req.query.interval_end) : Number(new Date())
    console.log(start)
    console.log(end)
    if (end < start || start < 0) {
        streamsDebug('[ERROR] Interval is wrong')
        return res.status(400).send({error: 'Bad interval defined'})
    }

    var before = new Date()
    const tmp = await values.aggregate([{
        $match:{
            device_ID: device.device_ID
            //$and: [{created_at: {$gte: start}},{created_at: {$lt: end}}]
        }
    },{
        $group:{
            _id: "$stream_name",
            values: {
                $push: {
                    timestamp: "$timestamp",
                    value: "$value",
                    longitude: "$longitude",
                    latitude: "$latitude"
                }
            }
        }
    }])
    console.log('este ')
    console.log(tmp)
    console.log('\n\n')
    var after = new Date()

    console.log(after-before)
    res.send(tmp)
})


module.exports = router